hail/src/main/scala/is/hail/annotations/RegionMemory.scala

```diff
@@ -0,0 +1,210 @@
+package is.hail.annotations
+
+import is.hail.utils._
+
+final class RegionMemory(pool: RegionPool) extends AutoCloseable {
+  private val usedBlocks = new ArrayBuilder[Long](4)
+  private val bigChunks = new ArrayBuilder[Long](4)
+
+  private var totalChunkMemory = 0L
+  private var currentBlock: Long = 0L
+  private var offsetWithinBlock: Long = _
+
+  // blockThreshold and blockByteSize are mutable because RegionMemory objects are reused with different sizes
+  var blockSize: Region.Size = -1
+  private var blockThreshold: Long = _
+  private var blockByteSize: Long = _
+
+  private val references = new ArrayBuilder[RegionMemory](4)
+  private var referenceCount: Long = _
+
+  def allocateNewBlock(): Unit = {
+    if (currentBlock != 0)
+      usedBlocks += currentBlock
+    currentBlock = pool.getBlock(blockSize)
+  }
+
+  def getCurrentBlock(): Long = currentBlock
+
+  private def allocateBigChunk(size: Long): Long = {
+    val o = pool.getChunk(size)
+    bigChunks += o
+    totalChunkMemory += size
+    o
+  }
+
+  def allocate(n: Long): Long = {
```

Comment:
I kind of feel like we could remove this signature of `allocate` all the way up. (its only use is in `appendByte` which is only used in some basic region tests.) Reduces the duplicated logic in this PR, but you don't have to do it here.